sw/device/lib/testing/dif_uart_test.cc

```diff
@@ -0,0 +1,819 @@
+                0b00000000'00000000'00000000'00000000);
+  EXPECT_WRITE32(UART_FIFO_CTRL_REG_OFFSET,
+                 0b00000000'00000000'00000000'00001100);
+
+  EXPECT_TRUE(dif_uart_watermark_rx_set(&dif_uart_, kDifUartWatermarkByte16));
+
+  EXPECT_READ32(UART_FIFO_CTRL_REG_OFFSET,
+                0b00000000'00000000'00000000'00000000);
+  EXPECT_WRITE32(UART_FIFO_CTRL_REG_OFFSET,
+                 0b00000000'00000000'00000000'00010000);
+
+  EXPECT_TRUE(dif_uart_watermark_rx_set(&dif_uart_, kDifUartWatermarkByte30));
+}
+
+// dif_uart_watermark_tx_set() tests
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInitialised, WatermarkTxSetFailureUartNull) {
+  EXPECT_FALSE(dif_uart_watermark_tx_set(nullptr, kDifUartWatermarkByte1));
+}
+
+/**
+ * Tests that all of the available TX watermark levels can be set.
+ */
+TEST_F(DifUartTestUartInitialised, WatermarkTxSetSuccess) {
+  EXPECT_READ32(UART_FIFO_CTRL_REG_OFFSET,
+                0b00000000'00000000'00000000'00000000);
+  EXPECT_WRITE32(UART_FIFO_CTRL_REG_OFFSET,
+                 0b00000000'00000000'00000000'00000000);
+
+  EXPECT_TRUE(dif_uart_watermark_tx_set(&dif_uart_, kDifUartWatermarkByte1));
+
+  EXPECT_READ32(UART_FIFO_CTRL_REG_OFFSET,
+                0b00000000'00000000'00000000'00000000);
+  EXPECT_WRITE32(UART_FIFO_CTRL_REG_OFFSET,
+                 0b00000000'00000000'00000000'00100000);
+
+  EXPECT_TRUE(dif_uart_watermark_tx_set(&dif_uart_, kDifUartWatermarkByte4));
+
+  EXPECT_READ32(UART_FIFO_CTRL_REG_OFFSET,
+                0b00000000'00000000'00000000'00000000);
+  EXPECT_WRITE32(UART_FIFO_CTRL_REG_OFFSET,
+                 0b00000000'00000000'00000000'01000000);
+
+  EXPECT_TRUE(dif_uart_watermark_tx_set(&dif_uart_, kDifUartWatermarkByte8));
+
+  EXPECT_READ32(UART_FIFO_CTRL_REG_OFFSET,
+                0b00000000'00000000'00000000'00000000);
+  EXPECT_WRITE32(UART_FIFO_CTRL_REG_OFFSET,
+                 0b00000000'00000000'00000000'01100000);
+
+  EXPECT_TRUE(dif_uart_watermark_tx_set(&dif_uart_, kDifUartWatermarkByte16));
+}
+
+// dif_uart_bytes_send() tests
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestBytesSendReceive, BytesSendFailureUartNull) {
+  EXPECT_FALSE(dif_uart_bytes_send(nullptr, bytes_send_data_, 1, nullptr));
+}
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestBytesSendReceive, BytesSendFailureDataNull) {
+  EXPECT_FALSE(dif_uart_bytes_send(&dif_uart_, nullptr, 1, nullptr));
+}
+
+/**
+ * Tests that the size of TX FIFO of bytes are written, when the FIFO is
+ * empty.
+ *
+ * Note: it is theoretically possible to write more than MAX FIFO size in case
+ *       the UART HW manages to send data out fast enough. This is a perfectly
+ *       legal behaviour, however is unlikely to happen.
+ */
+TEST_F(DifUartTestBytesSendReceive, TxFifoEmptySuccess) {
+  for (int i = 0; i < 32; ++i) {
+    EXPECT_READ32(UART_STATUS_REG_OFFSET,
+                  0b00000000'00000000'00000000'00000000);
+    EXPECT_WRITE32(UART_WDATA_REG_OFFSET, (uint32_t)bytes_send_data_[i]);
```

Comment:
Are you sure you want to sign-extend a byte into a word? It might be worthwhile to make `bytes_send_data` contain words rather than bytes.

Also, do not use C-style casts in C++; prefer `static_cast`.